torchrl/objectives/costs/dreamer.py

```diff
@@ -0,0 +1,268 @@
+        gamma (float, optional): the gamma discount factor. default: 0.99.
+        lmbda (float, optional): the lambda discount factor factor. default: 0.95
+        discount_loss (bool, optional): if True, the loss is discounted with a
+            gamma discount factor. default: False.
+    """
+
+    def __init__(
+        self,
+        actor_model: TensorDictModule,
+        value_model: TensorDictModule,
+        model_based_env: DreamerEnv,
+        imagination_horizon: int = 15,
+        gamma: int = 0.99,
+        lmbda: int = 0.95,
+        discount_loss: bool = False,  # for consistency with paper
+    ):
+        super().__init__()
+        self.actor_model = actor_model
+        self.value_model = value_model
+        self.model_based_env = model_based_env
+        self.imagination_horizon = imagination_horizon
+        self.gamma = gamma
+        self.lmbda = lmbda
+        self.discount_loss = discount_loss
+
+    def forward(self, tensordict: TensorDict) -> torch.Tensor:
+        with torch.no_grad():
+            tensordict = tensordict.select("state", "belief")
+            tensordict = tensordict.reshape(-1)  # 50 x 50 -> 2500
+
+        with hold_out_net(self.model_based_env), set_exploration_mode("random"):
+            tensordict = self.model_based_env.rollout(
+                max_steps=self.imagination_horizon,
+                policy=self.actor_model,
+                auto_reset=False,
+                tensordict=tensordict,
+            )
+
+            next_tensordict = step_mdp(
+                tensordict,
+                keep_other=True,
+            )
+            with hold_out_net(self.value_model):
+                next_tensordict = self.value_model(next_tensordict)
+
+        reward = tensordict.get("reward")
+        next_value = next_tensordict.get("state_value")
+        lambda_target = self.lambda_target(reward, next_value)
+        tensordict.set("lambda_target", lambda_target.detach())
+
+        if self.discount_loss:
+            discount = self.gamma * torch.ones_like(
+                lambda_target, device=tensordict.device
+            )
+            discount[:, 0] = 1
+            discount = discount.cumprod(dim=1).detach()
+            actor_loss = -(lambda_target * discount).sum((-2, -1)).mean()
+        else:
+            actor_loss = -lambda_target.sum((-2, -1)).mean()
+        loss_tensordict = TensorDict({"loss_actor": actor_loss}, [])
+        return loss_tensordict, tensordict.detach()
+
+    def lambda_target(self, reward: torch.Tensor, value: torch.Tensor) -> torch.Tensor:
+        done = torch.zeros(reward.shape, dtype=torch.bool, device=reward.device)
+        return vec_td_lambda_return_estimate(
+            self.gamma, self.lmbda, value, reward, done
+        )
+
+
+class DreamerValueLoss(LossModule):
+    """Dreamer Value Loss
+
+    Computes the loss of the dreamer value model. The value loss is computed as the mean squared error between the predicted value and the lambda target.
```

Comment:
not necessarily L2 since your loss is an arg of the constructor
maybe we just want to talk about the distance loss or smth?